# Analysis of DNA sequencing

In this first step, we will analyze the quality and length of PacBio HiFi and Nanopore R10.4 sequencing data prior to assembly. PacBio HiFi and Nanopore R10.4 reads are both high-accuracy datasets. We will extract the *k*-mers from these reads to estimate features of the genome (size, heterozygosity, repetitive content, ploidy).

Goals:
* assess the quality and length of long-read datasets
* estimate genome characteristics using *k*-mer approaches

## Nanoplot 

The tool NanoPlot takes as input a read file as fastq, fasta, or bam (compressed or not). Quality metrics cannot be computed with a fasta file, as it does not contain quality values (only nucleotides). Plots generated by NanoPlot can show length distribution and quality distribution. PacBio HiFi reads generally have a length around 10-15 kb, and a quality over 30 and reaching much higher values. Nanopore R10.4 reads are not limited in size; depending on the quality of the extraction, they can range from 10 kb to 100+ kb (even 1+ Mb). Their accuracy has improved over Nanopore R9.4.1 reads, but is lower than PacBio HiFi reads (from 5 to 40). 

## NanoPlot on PacBio HiFi reads

[NanoPlot](https://github.com/wdecoster/NanoPlot)

![nanoplot_hifi](s1_pic/nanoplot_hifi.png)

```sh
NanoPlot --tsv_stats --fastq ju765.hifi.decont.fastq.gz
```

![nanoplot_hifi](s1_pic/nanoplot_hifi_result.png)

## NanoPlot on Nanopore reads

[NanoPlot](https://github.com/wdecoster/NanoPlot)

![nanoplot_hifi](s1_pic/nanoplot_ont.png)

```sh
NanoPlot --tsv_stats --fastq ju765.ont.decont.fastq.gz
```

![nanoplot_hifi](s1_pic/nanoplot_ont_result.png)

## GenomeScope on PacBio HiFi reads

[jellyfish](https://github.com/gmarcais/Jellyfish)
[GenomeScope2](https://github.com/tbenavi1/genomescope2.0)

![jellyfish_count](s1_pic/jellyfish_count_hifi.png) 

![jellyfish_hist](s1_pic/jellyfish_hist_hifi.png) 

![genomescope](s1_pic/genomescope_hifi.png) 

```sh
jellyfish count --mer-len 21 ju765.hifi_reads.decont.3Gb.fastq.gz
jellyfish histo -o jellyfish.histo mer_counts.jf
genomescope2 --input jellyfish.histo --output . --kmer_length 21
```

![genomescope_result](s1_pic/linear_plot.png)

Two peaks can be distinguished, the first one being heterozygous *k*-mers and the second one homozygous *k*-mers. The model is not fitting well due to a trail of *k*-mers leading to the first peak, and results in an incorrect genome size estimation (which should be around 50 Mb). This trail of *k*-mers is due to alternative alleles as we have used a pool of individuals (in culture) to generate this dataset.

## Smudgeplot on PacBio HiFi reads

[Smudgeplot](https://github.com/KamilSJaron/smudgeplot)

![smudgeplot](s1_pic/smudgeplot.png)

```sh
mkdir tmp_smudge
ls ju765.hifi_reads.decont.3Gb.fastq.gz > FILES
kmc -k21 -ci1 -cs10000 @FILES kmcdb tmp_smudge
kmc_tools transform kmcdb histogram kmcdb_k21.hist -cx10000

L=$(smudgeplot.py cutoff kmcdb_k21.hist L)
U=$(smudgeplot.py cutoff kmcdb_k21.hist U)
echo $L $U

kmc_tools transform kmcdb -ci"$L" -cx"$U" dump -s kmcdb_L"$L"_U"$U".dump
smudgeplot.py hetkmers -o kmcdb_L"$L"_U"$U" < kmcdb_L"$L"_U"$U".dump

smudgeplot.py plot kmcdb_L"$L"_U"$U"_coverages.tsv
```

![smudgeplot_result](s1_pic/smudgeplot_smudgeplot.png)
![smudgeplot_result_log10](s1_pic/smudgeplot_smudgeplot_log10.png)

The strong signal around 1/2 2N supports a prediction of a diploid genome.
